consensus/consensus.go

```diff
@@ -0,0 +1,125 @@
+package consensus
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/builder"
+	consensusDB "github.com/NethermindEth/juno/consensus/db"
+	"github.com/NethermindEth/juno/consensus/driver"
+	"github.com/NethermindEth/juno/consensus/p2p"
+	"github.com/NethermindEth/juno/consensus/p2p/config"
+	"github.com/NethermindEth/juno/consensus/proposal"
+	"github.com/NethermindEth/juno/consensus/proposer"
+	"github.com/NethermindEth/juno/consensus/starknet"
+	"github.com/NethermindEth/juno/consensus/tendermint"
+	"github.com/NethermindEth/juno/consensus/types"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/db"
+	"github.com/NethermindEth/juno/utils"
+	"github.com/NethermindEth/juno/vm"
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/sourcegraph/conc/pool"
+)
+
+type ConsensusServices struct {
+	Host           host.Host
+	Proposer       proposer.Proposer[starknet.Value, starknet.Hash]
+	P2P            p2p.P2P[starknet.Value, starknet.Hash, starknet.Address]
+	Driver         *driver.Driver[starknet.Value, starknet.Hash, starknet.Address]
+	CommitListener driver.CommitListener[starknet.Value, starknet.Hash, starknet.Address]
+}
+
+func Init(
+	logger *utils.ZapLogger,
+	database db.KeyValueStore,
+	blockchain *blockchain.Blockchain,
+	vm vm.VM,
+	nodeAddress *starknet.Address,
+	validators tendermint.Validators[starknet.Address],
+	timeoutFn driver.TimeoutFn,
+	hostAddress string,
+	hostPrivateKey crypto.PrivKey,
+) (ConsensusServices, error) {
+	chainHeight, err := blockchain.Height()
+	if err != nil && !errors.Is(err, db.ErrKeyNotFound) {
+		return ConsensusServices{}, err
+	}
+	currentHeight := types.Height(chainHeight + 1)
+
+	tendermintDB := consensusDB.NewTendermintDB[starknet.Value, starknet.Hash, starknet.Address](database, currentHeight)
+
+	executor := builder.NewExecutor(blockchain, vm, logger, false, true) // TODO: We're currently skipping signature validation
+	builder := builder.New(blockchain, executor)
+
+	proposalStore := proposal.ProposalStore[starknet.Hash]{}
+	proposer := proposer.New(logger, &builder, &proposalStore, *nodeAddress, toValue)
+	stateMachine := tendermint.New(tendermintDB, logger, *nodeAddress, proposer, validators, currentHeight)
```

Comment:
Should nodeAddress be passed by reference here as well? In both `proposer.New` and `tendermint.New`